File: sdks/typescript/examples/test-large-files.ts
/**
 * Large File Streaming Test
 *
 * Tests:
 *   1. Write a 100MB file via SDK, read back and verify size
 *   2. Download 100MB file via signed URL, verify content integrity (SHA-256)
 *   3. Upload 100MB file via signed URL, verify content integrity
 *   4. Write 50MB via writeStream, read back via readStream
 *
 * Usage:
 *   npx tsx examples/test-large-files.ts
 */

import { createHash } from "crypto";
import { Sandbox } from "../src/index";

function green(msg: string) { console.log(`\x1b[32m✓ ${msg}\x1b[0m`); }
function red(msg: string) { console.log(`\x1b[31m✗ ${msg}\x1b[0m`); }
function bold(msg: string) { console.log(`\x1b[1m${msg}\x1b[0m`); }
function dim(msg: string) { console.log(`\x1b[2m  ${msg}\x1b[0m`); }

let passed = 0;
let failed = 0;

function check(desc: string, condition: boolean, detail?: string) {
  if (condition) {
    green(desc);
    passed++;
  } else {
    red(`${desc}${detail ? ` (${detail})` : ""}`);
    failed++;
  }
}

/** Generate deterministic data: repeating pattern so we can verify integrity */
function generateData(sizeMB: number): Uint8Array {
  const size = sizeMB * 1024 * 1024;
  const buf = new Uint8Array(size);
  // Fill with repeating 256-byte pattern
  for (let i = 0; i < size; i++) {
    buf[i] = i % 256;
  }
  return buf;
}

function sha256(data: Uint8Array): string {
  return createHash("sha256").update(data).digest("hex");
}

async function streamToBuffer(stream: ReadableStream<Uint8Array>): Promise<Uint8Array> {
  const reader = stream.getReader();
  const chunks: Uint8Array[] = [];
  let totalLen = 0;
  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    chunks.push(value);
    totalLen += value.length;
  }
  const result = new Uint8Array(totalLen);
  let offset = 0;
  for (const chunk of chunks) {
    result.set(chunk, offset);
    offset += chunk.length;
  }
  return result;
}

async function main() {
  bold("\n╔══════════════════════════════════════════════════╗");
  bold("║         Large File Streaming Test                ║");
  bold("╚══════════════════════════════════════════════════╝\n");

  let sandbox: Sandbox | null = null;

  try {
    sandbox = await Sandbox.create({ template: "base", timeout: 300 });
    green(`Created sandbox: ${sandbox.sandboxId}`);
    console.log();

    // ── Test 1: Write and read 100MB file via SDK ────────────────
    bold("━━━ Test 1: Write/read 100MB file via SDK ━━━\n");

    const data100 = generateData(100);
    const hash100 = sha256(data100);
    dim(`Generated 100MB data, SHA-256: ${hash100.slice(0, 16)}...`);

    const t0 = Date.now();
    await sandbox.files.write("/root/large100.bin", data100);
    const writeMs = Date.now() - t0;
    dim(`Write took ${writeMs}ms (${(100 * 1000 / writeMs).toFixed(1)} MB/s)`);
    green("100MB write completed");

    const t1 = Date.now();
    const readBack = await sandbox.files.readBytes("/root/large100.bin");
    const readMs = Date.now() - t1;
    dim(`Read took ${readMs}ms (${(100 * 1000 / readMs).toFixed(1)} MB/s)`);

    check("Read returns correct size", readBack.length === data100.length,
      `got ${readBack.length}, expected ${data100.length}`);
    const readHash = sha256(readBack);
    check("SHA-256 matches after read", readHash === hash100,
      `got ${readHash.slice(0, 16)}...`);
    console.log();

    // ── Test 2: Download 100MB via signed URL ────────────────────
    bold("━━━ Test 2: Download 100MB via signed URL ━━━\n");

    const dlUrl = await sandbox.downloadUrl("/root/large100.bin", { expiresIn: 300 });
    dim(`Download URL generated`);

    const t2 = Date.now();
    const dlRes = await fetch(dlUrl);
    check("Signed URL returns 200", dlRes.status === 200);

    const contentLength = dlRes.headers.get("content-length");
    dim(`Content-Length: ${contentLength}`);
    check("Content-Length is 100MB", contentLength === String(data100.length));

    const dlData = new Uint8Array(await dlRes.arrayBuffer());
    const dlMs = Date.now() - t2;
    dim(`Download took ${dlMs}ms (${(100 * 1000 / dlMs).toFixed(1)} MB/s)`);

    check("Downloaded size correct", dlData.length === data100.length);
    const dlHash = sha256(dlData);
    check("SHA-256 matches via signed URL download", dlHash === hash100);
    console.log();

    // ── Test 3: Upload 100MB via signed URL ──────────────────────
    bold("━━━ Test 3: Upload 100MB via signed URL ━━━\n");

    const upUrl = await sandbox.uploadUrl("/root/uploaded100.bin");
    dim(`Upload URL generated`);

    const t3 = Date.now();
    const upRes = await fetch(upUrl, {
      method: "PUT",
      body: data100,
    });
    const upMs = Date.now() - t3;
    dim(`Upload took ${upMs}ms (${(100 * 1000 / upMs).toFixed(1)} MB/s)`);
    check("Upload returns 204", upRes.status === 204);

    // Verify via SDK readBytes
    const upReadBack = await sandbox.files.readBytes("/root/uploaded100.bin");
    check("Uploaded file size correct", upReadBack.length === data100.length);
    const upHash = sha256(upReadBack);
    check("SHA-256 matches after signed URL upload", upHash === hash100);
    console.log();

    // ── Test 4: readStream / writeStream (50MB) ──────────────────
    bold("━━━ Test 4: readStream / writeStream 50MB ━━━\n");

    const data50 = generateData(50);
    const hash50 = sha256(data50);
    dim(`Generated 50MB data, SHA-256: ${hash50.slice(0, 16)}...`);

    // writeStream
    const t4 = Date.now();
    await sandbox.files.writeStream("/root/stream50.bin", data50);
    const wsMs = Date.now() - t4;
    dim(`writeStream took ${wsMs}ms`);
    green("50MB writeStream completed");

    // readStream
    const t5 = Date.now();
    const readStream = await sandbox.files.readStream("/root/stream50.bin");
    const streamData = await streamToBuffer(readStream);
    const rsMs = Date.now() - t5;
    dim(`readStream took ${rsMs}ms`);

    check("readStream returns correct size", streamData.length === data50.length,
      `got ${streamData.length}, expected ${data50.length}`);
    const streamHash = sha256(streamData);
    check("SHA-256 matches via readStream", streamHash === hash50);
    console.log();

  } catch (err: any) {
    red(`Fatal error: ${err.message}`);
    if (err.stack) dim(err.stack);
    failed++;
  } finally {
    if (sandbox) {
      await sandbox.kill();
      green("Sandbox killed");
    }
  }

  // --- Summary ---
  bold("========================================");
  bold(` Results: ${passed} passed, ${failed} failed`);
  bold("========================================\n");
  if (failed > 0) process.exit(1);
}

main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});

File: sdks/python/examples/test_large_files.py
#!/usr/bin/env python3
"""
Large File Streaming Test

Tests:
  1. Write/read 100MB file via SDK
  2. Download 100MB file via signed URL
  3. Upload 100MB file via signed URL
  4. read_stream / write_stream 50MB

Usage:
  python examples/test_large_files.py
"""

import asyncio
import hashlib
import sys
import time

import httpx

from opencomputer import Sandbox

GREEN = "\033[32m"
RED = "\033[31m"
BOLD = "\033[1m"
DIM = "\033[2m"
RESET = "\033[0m"

passed = 0
failed = 0


def green(msg: str) -> None:
    print(f"{GREEN}✓ {msg}{RESET}")


def red(msg: str) -> None:
    print(f"{RED}✗ {msg}{RESET}")


def bold(msg: str) -> None:
    print(f"{BOLD}{msg}{RESET}")


def dim(msg: str) -> None:
    print(f"{DIM}  {msg}{RESET}")


def check(desc: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        green(desc)
        passed += 1
    else:
        red(f"{desc} ({detail})" if detail else desc)
        failed += 1


def generate_data(size_mb: int) -> bytes:
    """Generate deterministic data: repeating 256-byte pattern."""
    pattern = bytes(range(256))
    repeats = (size_mb * 1024 * 1024) // 256
    return pattern * repeats


def sha256(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


async def collect_stream(stream) -> bytes:
    """Collect an async byte iterator into a single bytes object."""
    chunks = []
    async for chunk in stream:
        chunks.append(chunk)
    return b"".join(chunks)


async def byte_chunks(data: bytes, chunk_size: int = 256 * 1024):
    """Yield data in chunks as an async iterator."""
    for i in range(0, len(data), chunk_size):
        yield data[i : i + chunk_size]


async def main() -> None:
    global passed, failed

    bold("\n╔══════════════════════════════════════════════════╗")
    bold("║       Large File Streaming Test (Python)         ║")
    bold("╚══════════════════════════════════════════════════╝\n")

    sandbox = None

    try:
        sandbox = await Sandbox.create(template="base", timeout=300)
        green(f"Created sandbox: {sandbox.sandbox_id}")
        print()

        # ── Test 1: Write/read 100MB ─────────────────────────────
        bold("━━━ Test 1: Write/read 100MB file via SDK ━━━\n")

        data100 = generate_data(100)
        hash100 = sha256(data100)
        dim(f"Generated 100MB data, SHA-256: {hash100[:16]}...")

        t0 = time.time()
        await sandbox.files.write("/root/large100.bin", data100)
        write_s = time.time() - t0
        dim(f"Write took {write_s:.1f}s ({100 / write_s:.1f} MB/s)")
        green("100MB write completed")

        t1 = time.time()
        read_back = await sandbox.files.read_bytes("/root/large100.bin")
        read_s = time.time() - t1
        dim(f"Read took {read_s:.1f}s ({100 / read_s:.1f} MB/s)")

        check(
            "Read returns correct size",
            len(read_back) == len(data100),
            f"got {len(read_back)}, expected {len(data100)}",
        )
        read_hash = sha256(read_back)
        check("SHA-256 matches after read", read_hash == hash100)
        print()

        # ── Test 2: Download 100MB via signed URL ────────────────
        bold("━━━ Test 2: Download 100MB via signed URL ━━━\n")

        dl_url = await sandbox.download_url("/root/large100.bin", expires_in=300)
        dim("Download URL generated")

        t2 = time.time()
        async with httpx.AsyncClient(timeout=120.0) as http:
            dl_resp = await http.get(dl_url)
        check("Signed URL returns 200", dl_resp.status_code == 200)

        content_length = dl_resp.headers.get("content-length")
        dim(f"Content-Length: {content_length}")
        check("Content-Length is 100MB", content_length == str(len(data100)))

        dl_data = dl_resp.content
        dl_s = time.time() - t2
        dim(f"Download took {dl_s:.1f}s ({100 / dl_s:.1f} MB/s)")

        check("Downloaded size correct", len(dl_data) == len(data100))
        dl_hash = sha256(dl_data)
        check("SHA-256 matches via signed URL download", dl_hash == hash100)
        print()

        # ── Test 3: Upload 100MB via signed URL ──────────────────
        bold("━━━ Test 3: Upload 100MB via signed URL ━━━\n")

        up_url = await sandbox.upload_url("/root/uploaded100.bin")
        dim("Upload URL generated")

        t3 = time.time()
        async with httpx.AsyncClient(timeout=120.0) as http:
            up_resp = await http.put(up_url, content=data100)
        up_s = time.time() - t3
        dim(f"Upload took {up_s:.1f}s ({100 / up_s:.1f} MB/s)")
        check("Upload returns 204", up_resp.status_code == 204)

        up_read_back = await sandbox.files.read_bytes("/root/uploaded100.bin")
        check("Uploaded file size correct", len(up_read_back) == len(data100))
        up_hash = sha256(up_read_back)
        check("SHA-256 matches after signed URL upload", up_hash == hash100)
        print()

        # ── Test 4: read_stream / write_stream (50MB) ────────────
        bold("━━━ Test 4: read_stream / write_stream 50MB ━━━\n")

        data50 = generate_data(50)
        hash50 = sha256(data50)
        dim(f"Generated 50MB data, SHA-256: {hash50[:16]}...")

        # write_stream
        t4 = time.time()
        await sandbox.files.write_stream("/root/stream50.bin", byte_chunks(data50))
        ws_s = time.time() - t4
        dim(f"write_stream took {ws_s:.1f}s")
        green("50MB write_stream completed")

        # read_stream
        t5 = time.time()
        stream = await sandbox.files.read_stream("/root/stream50.bin")
        stream_data = await collect_stream(stream)
        rs_s = time.time() - t5
        dim(f"read_stream took {rs_s:.1f}s")

        check(
            "read_stream returns correct size",
            len(stream_data) == len(data50),
            f"got {len(stream_data)}, expected {len(data50)}",
        )
        stream_hash = sha256(stream_data)
        check("SHA-256 matches via read_stream", stream_hash == hash50)
        print()

    except Exception as e:
        red(f"Fatal error: {e}")
        import traceback

        traceback.print_exc()
        failed += 1
    finally:
        if sandbox:
            await sandbox.kill()
            green("Sandbox killed")

    # --- Summary ---
    bold("========================================")
    bold(f" Results: {passed} passed, {failed} failed")
    bold("========================================\n")
    if failed > 0:
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())
